src/librustc/ty/layout.rs

```diff
@@ -511,41 +512,70 @@ pub struct Struct {
     /// If true, the size is exact, otherwise it's only a lower bound.
     pub sized: bool,
 
-    /// Offsets for the first byte of each field.
+    /// Offsets for the first byte of each field, ordered in increasing order.
+    /// The ith element of this vector is not always the ith field of the struct.
     /// FIXME(eddyb) use small vector optimization for the common case.
     pub offsets: Vec<Size>,
 
+    /// Maps field indices to indices in the offsets Vec.
+    /// The offset of the ith field is offsets[field_index[i]].
+    /// FIXME (camlorn) also consider small vector  optimization here.
+    pub field_index: Vec<u32>,
+
     pub min_size: Size,
 }
 
 impl<'a, 'gcx, 'tcx> Struct {
-    pub fn new(dl: &TargetDataLayout, packed: bool) -> Struct {
-        Struct {
+    pub fn new<I>(dl: &TargetDataLayout, fields: I,
+                  repr: attr::ReprAttr, is_enum_variant: bool,
+                  scapegoat: Ty<'gcx>) -> Result<Struct, LayoutError<'gcx>>
+        where I: Iterator<Item=Result<&'a Layout, LayoutError<'gcx>>>{
+            let packed = repr == attr::ReprAny;
+        let mut ret = Struct {
             align: if packed { dl.i8_align } else { dl.aggregate_align },
             packed: packed,
             sized: true,
             offsets: vec![],
+            field_index: vec![],
             min_size: Size::from_bytes(0),
-        }
+        };
+        ret.fill_in_fields(dl, fields, scapegoat, repr, is_enum_variant)?;
+        Ok(ret)
     }
 
-    /// Extend the Struct with more fields.
-    pub fn extend<I>(&mut self, dl: &TargetDataLayout,
+    fn fill_in_fields<I>(&mut self, dl: &TargetDataLayout,
```

Comment:
You can probably move the whole body of `fill_in_fields` into `fn new`. Also maybe rename it to `from_fields`.

**EDIT**: Oh is it because of `if fields.len() == 0 {return Ok(())};` below?